Implement plugin architecture for new instruments (#168)
HiPERCAM/hipercam: docs/plugins.rst

.. Developer guide for third-party instrument plugins

.. include:: globals.rst

Writing an instrument plugin
****************************

The |hiper| pipeline supports third-party instruments through a plugin system
based on Python entry-points.  A plugin is a Python module that exposes a
small set of classes and a helper function, which the pipeline uses to read raw

> data and present it as a stream of :class:`~hipercam.MCCD` objects.

Reviewer: These auto hyperlinks don't actually work at the moment, since we don't build the full autoapi. We can enable that (there's the `autoapi_generate_api_docs` param in `config.py`), but it takes a while and there are loads of warnings which is why I didn't in #161. Could be part of a larger refresh of the docs.

This page explains how to write such a plugin.  As a concrete example
throughout, we assume an instrument called ``myinstrument`` that has a single
1024x1024 CCD and stores its data in a multi-extension FITS file: one primary
HDU (header only, carrying run-level metadata) followed by one image extension
per acquired frame.

Overview
========

The pipeline discovers plugins through the ``hipercam.instruments``
entry-point group.  When a data source string of the form

> ``'<instrument>:local'`` is passed to

Reviewer: Just to check: in the code you have `'<instrument>:local'` and ``'<instrument>:server'``, but in this doc you only mention `local`. Is `PluginServSpool` etc just there for a future use case for now?

Author: This should probably mention the server case as well. The docs however don't show an implementation for the server case since it requires the end-user to write their own version of the `FileServer` we use for HiPERCAM or ULTRACAM, and this is a complex task which depends on how the data is streamed to disc.

Reviewer: Agreed, I don't think we want to imply that users have to implement their own fileserver system and for most local should be fine. Maybe we could just mention somewhere in the docs that direct streaming is *possible* using the pipeline (and is implemented for HiPERCAM and ULTRACAM) but is not currently supported through the plugin architecture.

.. note::
   If you are using ``pip install -e .`` to install your package in 
   editable (development) mode, you must re-run ``pip install -e .`` 
   each time you change the entry-point declaration, even in editable 
   (development) mode, so that the entry-point metadata is regenerated.

In ``__init__.py`` you can put any module-level code you like, but you must 
at least import the module level objects ``Rdata``, ``get_ccd_pars`` and 
``supports_server_access`` which we will define in the ``example.py`` file::

    # src/myinst_plugin/__init__.py
    from .instrument import Rdata, get_ccd_pars, supports_server_access

Module contract
===============

Your module must expose the names listed in the overview table.  Below is the
minimum skeleton::

    # src/myinst_plugin/instrument.py

    from hipercam.instruments import IendError, RdataBase

    supports_server_access = False

    class Rdata(RdataBase):  # data reader — the main class
        ...

    def get_ccd_pars(resource, server=False):
        ...

Implementing Rdata
==================

``Rdata`` is the central class.  The pipeline expect an initialiser with 
the signature below::

>     Rdata(run, nframe, server, **kwargs)

Reviewer: Is there a reason you went with `run` here as the argument instead of `fname`? `hcam.Rdata` uses `fname`, `ucam.Rdata` uses `run` but you have the comment in `ucam.Rhead` that `fname` is preferred and `run` is only kept for backwards compatibility.  Especially as below you have `self._fname = run`, I'd think for plugin users it would just be easier to use `fname` from the start.

Author: good point - I've removed as many references to run from the docs as possible

            # get the header and data for the current frame.
            hdu   = self._hdul[self.nframe]  # 1-based: HDU 0 is the primary
            data  = hdu.data.astype(np.float32)
            fhead = hdu.header               # raw FITS header for this extension

            # delegate construction of the MCCD object to a helper method.
            mccd = self._build_mccd(data, fhead, self.nframe)

            # increment the frame counter so a subsequent call
            # (with nframe=None) will read the next frame in sequence.
            self.nframe += 1

            # return the MCCD object to the caller
            return mccd

Constructing MCCD objects
=========================

The :class:`~hipercam.MCCD` is the standard data container used throughout the
pipeline.  A :class:`~hipercam.MCCD` consists of a collection of one or more
:class:`~hipercam.CCD` objects, each of which contains one or more 
:class:`~hipercam.Window` objects. 

A :class:`~hipercam.CCD` can contain multiple windows to deal with instruments
that have more than one output amplifier, each reading out a different region 
of the CCD.

It is assembled from the inside out::

>     `~hipercam.header.Header`  →  Winhead  →  Window  →  CCD  →  MCCD

Reviewer: The link doesn't work when indented, so just remove:
```suggestion
    Header  →  Winhead  →  Window  →  CCD  →  MCCD
```

Author: Done

Top-level header
----------------

The top-level header of the :class:`~hipercam.MCCD` is an

> ``hipercam.header.Header`` or ``~astropy.io.fits.Header`` object.  

Reviewer: Include links:
```suggestion
The top-level header of the :class:`~hipercam.MCCD` is a
:class:`hipercam.Header` object, which itself is a lightweight version of an
`Astropy FITS Header <https://docs.astropy.org/en/stable/io/fits/usage/headers.html>`.
```

The pipeline requires the following keywords to be present:

=============  ======  =======================================================
Keyword        Type    Description
=============  ======  =======================================================
``INSTRUME``   str     Instrument name
``TIMSTAMP``   str     UTC timestamp of the mid-exposure in ISO 8601 format,
                       e.g. ``'2026-04-16T21:30:00.000000000'``
``MJDINT``     int     Integer part of the UTC mid-exposure MJD
``MJDFRAC``    float   Fractional part of the UTC mid-exposure MJD

> ``MJDUTC``     float   Full MJD(UTC), equal to ``MJDINT + MJDFRAC``

Reviewer: Could these not be derived from the `TIMESTAMP`? I assume it's all for speed, and there's a good reason INT and FRAC are stored separately. Incidentally the only actual use of `MJDUTC` in the pipeline is in `shiftadd`, everywhere else just does `mjdint+mjdfrac`.

Author: Not just speed, but also retaining precision on timing data. The TIMSTAMP entry is really for convenience, proper timing is done with the MJDINT and MJDFRAC data

> ``GOODTIME``   bool    ``True`` if the timestamp is reliable

Reviewer: This is minor, but `GOODTIME` feels quite specific to include as a mandatory card - could we make it optional and assume `True` if missing? I don't think it would even need any changes in the reduction code, every reference I could find uses `ccd.head.get("GOODTIME", True)` so if it's missing it shouldn't break anything.

Author: I think it's important to keep as good practice. Even it is `True` by default you want to make sure that people think about whether a timestamp is reliable IMO

``EXPTIME``    float   Exposure duration in seconds
``NFRAME``     int     Frame number within the run (1-based)
=============  ======  =======================================================

In addition to this, any other FITS keywords may be included as desired. 

Example::

>     thead = fits.Header()

Reviewer: Since the other examples include imports:
```suggestion
    from astropy.io import fits
    thead = fits.Header()
```

    thead['INSTRUME'] = ('myinstrument', 'Instrument name')
    thead['TIMSTAMP'] = (fhead['DATE-OBS'], 'UTC timestamp')
    mjd     = fhead['MJD-OBS']
    mjdint  = int(mjd)
    mjdfrac = mjd - mjdint
    thead['MJDINT']   = (mjdint,  'Integer part of MJD(UTC)')
    thead['MJDFRAC']  = (mjdfrac, 'Fractional part of MJD(UTC)')
    thead['MJDUTC']   = (mjd,     'MJD(UTC)')
    thead['GOODTIME'] = (True,    'Timestamp reliable')
    thead['EXPTIME']  = (float(fhead['EXPTIME']), 'Exposure time (s)')
    thead['NFRAME']   = (nframe,  'Frame number')

Per-CCD header
--------------

There is no separate header object for a :class:`~hipercam.CCD`.  Instead,
CCD-level information — in particular per-CCD timing data — is stored in the

> header of the **first** :class:`~hipercam.Window` in that CCD.  This header is

Reviewer: More out of curiosity, what was the reasoning behind doing this and not storing the timing header in the CCD? Too late to change now, but maybe worth a sentence to explain given this will be for people not used to the data format.

Author: I have absolutely no idea - this was a design decision Tom made way back in the day. We could just write "for historical reasons". It's not a good idea to change now since it would mean supporting old and new `.hcm` file formats

Reviewer: To be clear, I wasn't suggesting actually changing how the files were written. But we could allow adding a header to the CCD class in the API, which would actually get written to the window header when reading/writing files. Not really a priority, just a thought.

a :class:`hipercam.Header` object (a lightweight FITS-compatible header class)

> and must carry the same timing keywords as the top-level header:

Reviewer: Just trying to get this straight in my head: the `hipercam.header.Header` class says it "Simulates the basic functionality of astropy.io.fits.Header objects while being lighter weight and thus more efficient." So in general we prefer to use our class since it's quicker (by not doing extra checks etc) but the API is the same as the Astropy class, so someone could use the Astropy one instead. At least that's what's implied for the MCCD header, here we only refer to the pipeline object. I guess that's for speed, since one MCCD header is nothing compared to the huge number of per-frame headers? Could be worth clarifying.

Author: Yes - this is for speed. We read or write many `hipercam.header.Header` from a file, but only one MCCD level header. So we prefer our own class here. Ive added some clarifying text

=============  ===============================================================
Keyword        Description
=============  ===============================================================
``TIMSTAMP``   UTC timestamp (ISO 8601)
``MJDINT``     Integer part of MJD(UTC)
``MJDFRAC``    Fractional part of MJD(UTC)
``MJDUTC``     Full MJD(UTC)
``GOODTIME``   Timestamp reliability flag
``EXPTIME``    Exposure time (s)

> ``NFRAME``     Frame number

Reviewer: Same point as above about which of these keywords could be optional. For instance if you didn't have per-CCD timing (or I guess only a single CCD) then could you leave the timing cards here blank and just inherit the MCCD time? 
Also, would there ever be a reason to expect `NFRAME` in a CCD header to not be the same as `NFRAME` in the MCCD header? I think all the references in the pipeline use `mccd.head["NFRAME"]`.

Author: I don't know about making changes here. I don't think this PR is the place to change the data format of the pipeline

Reviewer: Again apologies, I wasn't actually suggesting changing the data format or output files. Just thinking about ways we could reduce the amount the user needed to input.

    ccd_head['NFRAME']   = (nframe,  'Frame number')

Building Winhead and Window
---------------------------

A :class:`~hipercam.Winhead` describes the position and binning of a
rectangular readout region of the CCD. Windows may only take up a subset
of the full CCD area, and may be binned. For a full-frame 1024x1024 readout
with 1x1 binning::

    from hipercam.window import Winhead, Window

    # Winhead(llx, lly, nx, ny, xbin, ybin, outamp, head=ccd_head)
    #
    #   llx, lly  — lower-left corner in **unbinned pixels**, 1-based
    #   nx, ny    — window dimensions in **binned** pixels
    #   xbin,ybin — binning factors
    #   outamp    — output amplifier location; 'UL', 'UR', 'LL' etc. '' means unknown
    #   head      — per-CCD header created above, attached to
    #               the first (and here only) window
    #
    winhd = Winhead(1, 1, 1024, 1024, 1, 1, '', head=ccd_head)
    win   = Window(winhd, data=data)

.. note::

   If a CCD has more than one readout window, you need to create
   a :class:`~hipercam.Window` object for each one. Only the **first** 
   window in the :class:`~hipercam.CCD` carries the per-CCD timing header; 

>    all subsequent windows should omit the `head` parameter.

Reviewer: Does it actually break if you have headers on more than the first? What if you have per-Window header information, e.g. different gain or readout values? Again trying to think from the PoV of someone with an existing different data format.

Author: No I don't think it breaks, but the pipeline gets all the data from the first one IIRC

Building CCD and MCCD
---------------------

A :class:`~hipercam.CCD` is assembled from one or more :class:`~hipercam.Window` 
objects. The :class:`~hipercam.Group` object is a simple wrapper around
an ``OrderedDict`` that we can use to create a group of windows::

    from hipercam import Group, CCD

    # CCD with total size 1024x1024, no pre/over-scan
    ccd_obj = CCD(Group(Window), nxtot=1024, nytot=1024, nxpad=0, nypad=0)
    # add the single window to the CCD with label '1'

>     ccd['1'] = win

Reviewer: First, `ccd_obj` should just be `ccd`, or vice versa? But also it's maybe easier to explicitly make the group as you do for `ccds` below, since it's not immediately obvious that you pass the *class* to the group first, then create the CCD, then add the window instance.

```suggestion
    # create an empty Window Group
    winds = Group(Window)
    # CCD with total size 1024x1024, no pre/over-scan
    ccd = CCD(winds, nxtot=1024, nytot=1024, nxpad=0, nypad=0)
    # add the single window to the CCD with label '1'
    ccd['1'] = win
```

Or actually you could mirror below and add the window first?
```suggestion
    winds = Group(Window)
    # add the single window to the group with label '1'
    winds['1'] = win
    # CCD with total size 1024x1024, no pre/over-scan
    ccd = CCD(winds, nxtot=1024, nytot=1024, nxpad=0, nypad=0)
```

For what it's worth, I feel this is a really awkward API. From reading the `Group` docs, maybe this would also work:
```python
winds = Group(Window, '1'=win)
```
although having to pass the ftype to the Group seems unnecessary, over it just checking the values.
Really it would be far easier for a user to just pass a simple dict of label/window pairs to the (M)CCD and not worry about the Group class at all (it would still make one internally and raise errors for non-string keys or different value classes)
```python
winds = {'1': win}
ccd = CCD(winds, nxtot=1024, nytot=1024, nxpad=0, nypad=0)
ccds = {'1': ccd}
mccd = MCCD(ccds, head=thead)
```
But this is all probably for a different issue.

Author: I've gone with the first suggestion. 

The advantage of the `Group` class is that it enforces all members are the correct class, no? A bit like numpy arrays vs lists?

I like the idea of allowing the user to pass in a simple dictionary - we could raise that in a separate issue, which would also update these docs

Reviewer: I had a little play around with Group, but this is a very back-end thing. I agree it's a different issue.

Finally we can create a group of one or more CCDs and use it to construct the
:class:`~hipercam.MCCD` by providing the group and the top-level header::

    ccds = Group(CCD)
    # add the single CCD to the group with label '1'

>     ccds['1'] = ccd_obj

Reviewer: ```suggestion
    ccds['1'] = ccd
```

        thead['TIMSTAMP'] = (tstamp,   'UTC timestamp')
        thead['MJDINT']   = (mjdint,   'Integer part of MJD(UTC)')
        thead['MJDFRAC']  = (mjdfrac,  'Fractional part of MJD(UTC)')
        thead['MJDUTC']   = (mjd,      'MJD(UTC)')
        thead['GOODTIME'] = (True,     'Timestamp reliable')
        thead['EXPTIME']  = (exptime,  'Exposure time (s)')
        thead['NFRAME']   = (nframe,   'Frame number')

        # --- per-CCD header, stored in the first window (hipercam.Header) ---
        ccd_head = Header()
        ccd_head['TIMSTAMP'] = (tstamp,  'UTC timestamp')
        ccd_head['MJDINT']   = (mjdint,  'Integer part of MJD(UTC)')
        ccd_head['MJDFRAC']  = (mjdfrac, 'Fractional part of MJD(UTC)')
        ccd_head['MJDUTC']   = (mjd,     'MJD(UTC)')
        ccd_head['GOODTIME'] = (True,    'Timestamp reliable')
        ccd_head['EXPTIME']  = (exptime, 'Exposure time (s)')
        ccd_head['NFRAME']   = (nframe,  'Frame number')

        # --- window, CCD, MCCD ---
        winhd   = Winhead(1, 1, 1024, 1024, 1, 1, '', head=ccd_head)
        win     = Window(winhd, data=data)

        # CCD with total size 1024x1024, no pre/over-scan

>         ccd_obj = CCD(Group(Window), nxtot=1024, nytot=1024, nxpad=0, nypad=0)

Reviewer: Any changes above, but at minimum correcting `ccd`/`ccd_obj`.

        # add the single window to the CCD with label '1'
        ccd['1'] = win

        ccds    = Group(CCD)
        ccds['1'] = ccd_obj
        return MCCD(ccds, head=thead)

Implementing get_ccd_pars
=========================

:func:`get_ccd_pars` is called by the pipeline to determine CCD geometry
before any frames are read (for example, to set up display windows).  It must

> return an :class:`~collections.OrderedDict` mapping each CCD label to a

Reviewer: Python dicts have been guaranteed to be ordered since 3.7. It doesn't seem we use anything unique to `collections.OrderedDict`, and while the package technically says it's compatible with Python3.6 and above I doubt that's the case (and do we really want to support a version that went EoL in 2021?).
In other words, I'd support just asking to return a simple dict here, again keeping things simple for the plugin user.

4-tuple ``(nxmax, nymax, nxpad, nypad)``:

==========  =====================================================================
Value       Meaning
==========  =====================================================================
``nxmax``   Full unbinned X dimension of the CCD imaging area
``nymax``   Full unbinned Y dimension of the CCD imaging area
``nxpad``   Extra X pixels to allow for pre/over-scan regions (0 if none)
``nypad``   Extra Y pixels to allow for pre/over-scan regions (0 if none)
==========  =====================================================================

For ``myinstrument`` with its single 1024x1024 CCD and no scan regions::

    from collections import OrderedDict

    def get_ccd_pars(resource, server=False):
        """Return CCD geometry for myinstrument.

        Arguments::

           resource : str
              Run name.  Not used; myinstrument has fixed geometry.

           server : bool

>               Unused; present for API compatibility.

Reviewer: See comments in `spooler.py`, but I don't really get why these would actually be needed? The only place we call the `hcam`/`ucam`/`mod.get_ccd_pars()` is in `spooler.get_ccd_pars()`, and then it's already behind an if block - I don't see why we'd need to pass unused arguments?
